### PR TITLE
Apply changes from the validator spec

### LINKED ISSUE #2034: v0.12.1 - State transition
### Description
Implement state transition changes for 0.12.1 spec (when it's out).

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Update the calculation used for subnet IDs.
Add requirement that Eth1Data votes must not reduce the number of deposits.

## Fixed Issue(s)
fixes #2034

### CODE REVIEW

File: networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/AttestationGossipManagerTest.java [hunk 1]
```diff
@@ -52,39 +52,43 @@
...
   public void onNewAttestation_afterMatchingAssignment() {
+    final Attestation attestation = dataStructureUtil.randomAttestation();
+    final Attestation attestation2 =
+        new Attestation(
+            dataStructureUtil.randomBitlist(),
+            dataStructureUtil.randomAttestationData(UnsignedLong.valueOf(13)),
+            dataStructureUtil.randomSignature());
+    final int subnetId = computeSubnetId(attestation);
+    // Sanity check the attestations are for the same subnet
+    assertThat(computeSubnetId(attestation2)).isEqualTo(subnetId);
```
Comment (cemozerr) [inline networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/AttestationGossipManagerTest.java:81]: I'm confused with how we can be sure that these random attestations have the same subnetId?

File: networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/AttestationTopicSubscriber.java [hunk 2]
```diff
@@ -24,22 +23,40 @@
...
     this.eth2Network = eth2Network;
+    this.recentChainData = recentChainData;
   }
 
   public synchronized void subscribeToCommitteeForAggregation(
       final int committeeIndex, final UnsignedLong aggregationSlot) {
-    final int subnetId = committeeIndexToSubnetId(committeeIndex);
+    recentChainData
+        .getBestState()
+        // Not point aggregating for historic slots and we can't calculate the subnet ID
```
Comment (cemozerr) [inline networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/AttestationTopicSubscriber.java]: nit:
```suggestion
        // No point aggregating for historic slots and we can't calculate the subnet ID
```

**Review by cemozerr — APPROVED** [review]

File: networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/AttestationTopicSubscriber.java [hunk 3]
```diff
@@ -24,22 +23,40 @@
...
+        .filter(state -> state.getSlot().compareTo(aggregationSlot) <= 0)
```
Comment (mbaxter) [inline networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/AttestationTopicSubscriber.java:50]: Not sure I understand this line - don't we set up subscriptions at the beginning of each epoch, so wouldn't we expect that the aggregationSlot will be in the future?  Should this be:
```suggestion
        .filter(state -> state.getSlot().compareTo(aggregationSlot) > 0)
```
?

Comment (ajsutton) [reply networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/AttestationTopicSubscriber.java:50]: Yes the aggregation slot should be greater than the current head slot (so it's in the future).  so `state.getSlot() <= aggregationSlot`.

File: networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/AttestationGossipManagerTest.java (hunk 1 above)
Comment (ajsutton) [reply networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/AttestationGossipManagerTest.java:81]: Remember that we use a fixed seed for all randomness, so they're completely repeatable.  And then I carefully selected the slot for the second attestation to make it line up.

**Pull request merged**